openaddr/util/__init__.py

```diff
@@ -205,3 +210,118 @@ def s3_key_url(key):
     path = join(key.bucket.name, key.name.lstrip('/'))
     
     return urljoin(base, path)
+
+def get_cpu_times():
+    ''' Return Linux CPU usage times in jiffies.
+    
+        See http://stackoverflow.com/questions/1420426/how-to-calculate-the-cpu-usage-of-a-process-by-pid-in-linux-from-c
+    '''
+    if not exists('/proc/stat') or not exists('/proc/{}/stat'.format(getpid())):
+        return None, None, None
+    
+    with open('/proc/stat') as file:
+        stat = re.split(r'\s+', next(file).strip())
+        time_total = sum([int(s) for s in stat[1:]])
+
+    with open('/proc/{}/stat'.format(getpid())) as file:
+        stat = next(file).strip().split(' ')
+        utime, stime = (int(s) for s in stat[13:15])
+    
+    return time_total, utime, stime
+
+def get_diskio_bytes():
+    ''' Return bytes read and written.
+    
+        See http://stackoverflow.com/questions/3633286/understanding-the-counters-in-proc-pid-io
+    '''
+    if not exists('/proc/{}/io'.format(getpid())):
+        return None, None
+    
+    read_bytes, write_bytes = None, None
+    
+    with open('/proc/{}/io'.format(getpid())) as file:
+        for line in file:
+            bytes = re.split(r':\s+', line.strip())
+            if 'read_bytes' in bytes:
+                read_bytes = int(bytes[1])
+            if 'write_bytes' in bytes:
+                write_bytes = int(bytes[1])
+    
+    return read_bytes, write_bytes
+
+def get_network_bytes():
+    ''' Return bytes sent and received.
+    '''
+    if not exists('/proc/net/netstat'):
+        return None, None
+    
+    sent_bytes, recv_bytes = None, None
+    
+    with open('/proc/net/netstat') as file:
+        for line in file:
+            columns = line.strip().split()
+            if 'IpExt:' in line:
+                values = next(file).strip().split()
+                netstat = {k: int(v) for (k, v) in zip(columns[1:], values[1:])}
+                sent_bytes, recv_bytes = netstat['OutOctets'], netstat['InOctets']
+    
+    return sent_bytes, recv_bytes
+
+def get_memory_usage():
```

Comment:
VMRSS is of interest here too; that's resident memory size. It will matter if a machine runs out of RAM.